### PR TITLE
[RTM when tests pass] Using integrate_extra_dynamics instead of duplicating code

### PR DESCRIPTION
**Pull request by EveCharbie**
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document [docs/contribution.md]?
* [x] Have you checked to ensure there aren't other open [Pull Requests] for the same update/change?
* [x] Have you opened/linked the issue related to your pull request?
* [x] Have you used the tag [WIP] for on-going changes, and removed it when the pull request was ready?
* [x] When ready to merge, have you sent a comment pinging @pariterre in it?

### New Feature Submissions:

1. [ ] Does your submission pass the tests (if not please explain why this is intended)?
2. [ ] Did you write a proper documentation (docstrings and ReadMe)
3. [ ] Have you linted your code locally prior to submission (using the command: `black . -l120 --exclude "external/*"`)?

### Changes to Core Features:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new examples for your core changes, as applicable?
* [ ] Have you written new tests for your core changes, as applicable?

- - -
This change is [Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/780)

### CODE REVIEW

**Comment by EveCharbie** [comment]
This branch was started on top of PR #765 
I cannot black while I am away.

**Review by pariterre — CHANGES_REQUESTED** [review]
> Reviewed 4 of 5 files at r1, 9 of 11 files at r2, 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/780)* status: all files reviewed, 1 unresolved discussion (waiting on @EveCharbie)

___
*[`bioptim/dynamics/ode_solver.py` line 89 at r3](https://reviewable.io/reviews/pyomeca/bioptim/780#-Nh2FHqz7I2kh6cMZHSc:-Nh2FHqz7I2kh6cMZHSd:bjlcql2) ([raw file](https://github.com/pyomeca/bioptim/blob/004671f923191aa23a19654c7ee148bc67682fd1/bioptim/dynamics/ode_solver.py#L89)):*
> ```Python
>         extra_dynamics = []
>         for i in range(1, len(nlp.dynamics_func)):
>             allow_free_variables = nlp.dynamics_func[i].has_free()
> ```

This can be set to True all the time

**Review by EveCharbie — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/780)* status: 13 of 15 files reviewed, 1 unresolved discussion (waiting on @pariterre)

___
*[`bioptim/dynamics/ode_solver.py` line 89 at r3](https://reviewable.io/reviews/pyomeca/bioptim/780#-Nh2FHqz7I2kh6cMZHSc:-Nh2Gc72A90PXto2aAAV:b-896fix) ([raw file](https://github.com/pyomeca/bioptim/blob/004671f923191aa23a19654c7ee148bc67682fd1/bioptim/dynamics/ode_solver.py#L89)):*
<details><summary><i>Previously, pariterre (Pariterre) wrote…</i></summary><blockquote>

This can be set to True all the time
</blockquote></details>

Done.

File: bioptim/dynamics/integrator.py
Comment (pariterre) [inline bioptim/dynamics/integrator.py:777]: Please investigate these

```python
xp_j = 0
for r in range(1, self.degree + 1):
	xp_j += self._c[r + 1 if r == 1 else r, j] * states[r]
```			
				
```python
xp_j = 0
for r in range(self.degree + 1):
	if r == 1:
		continue
	xp_j += self._c[r + 1 if r < 1 else r, j] * states[r]
```


```python
xp_j = self._c[0, j] * states[0]
for r in range(2, self.degree + 2):
	# We need to subtract 1 to r because we skipped it starting at 2
	xp_j += self._c[r - 1, j] * states[r]
```

Comment (EveCharbie) [reply bioptim/dynamics/integrator.py:777]: Done.

**Comment by EveCharbie** [comment]
@pariterre everything is good on my side, please wait for @Kev1CO's or Mickael's confirmation before merging.

**Review by pariterre — APPROVED** [review]
:lgtm:

> Reviewed 1 of 2 files at r4, 5 of 5 files at r5, 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/780#-:-NhC4iNoE8ECaPoM8TyY:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @EveCharbie)

**Review by pariterre — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pyomeca/bioptim/780#-:-NhRu2YxCdhLrAX9DAaG:bnfp4nl)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @EveCharbie)

**Pull request merged**
